tests/test_fakecraftcli_dispatcher.py

```diff
@@ -0,0 +1,468 @@
+            self._executed.append(parsed_args)
+
+    class MyCommand2(BaseCommand):
+        """Expected to not be instantiated, or parse args, or run."""
+
+        name = "command2"
+        help_msg = "some help"
+        overview = "fake overview"
+
+        def __init__(self, *args):
+            raise AssertionError
+
+        def fill_parser(self, parser):
+            raise AssertionError
+
+        def run(self, parsed_args):
+            raise AssertionError
+
+    groups = [CommandGroup("title", [MyCommand1, MyCommand2])]
+    dispatcher = Dispatcher("appname", groups)
+    dispatcher.pre_parse_args(["command1"])
+    dispatcher.load_command("config")
+    dispatcher.run()
+    assert isinstance(MyCommand1._executed[0], argparse.Namespace)
+
+
+def test_dispatcher_global_arguments_default():
+    """The dispatcher uses the default global arguments."""
+    cmd = create_command("somecommand")
+    groups = [CommandGroup("title", [cmd])]
+
+    dispatcher = Dispatcher("appname", groups)
+    assert dispatcher.global_arguments == _DEFAULT_GLOBAL_ARGS
+
+
+def test_dispatcher_global_arguments_extra_arguments():
+    """The dispatcher uses the default global arguments."""
+    cmd = create_command("somecommand")
+    groups = [CommandGroup("title", [cmd])]
+
+    extra_arg = GlobalArgument("other", "flag", "-o", "--other", "Other stuff")
+    dispatcher = Dispatcher("appname", groups, extra_global_args=[extra_arg])
+    assert dispatcher.global_arguments == _DEFAULT_GLOBAL_ARGS + [extra_arg]
+
+
+# --- Tests for the base command
+
+
+def test_basecommand_holds_the_indicated_info():
+    """BaseCommand subclasses ."""
+
+    class TestClass(BaseCommand):
+        help_msg = "help message"
+        name = "test"
+        overview = "fake overview"
+
+    config = "test config"
+    tc = TestClass(config)
+    assert tc.config == config
+
+
+def test_basecommand_fill_parser_optional():
+    """BaseCommand subclasses are allowed to not override fill_parser."""
+
+    class TestClass(BaseCommand):
+        help_msg = "help message"
+        name = "test"
+        overview = "fake overview"
+
+        def __init__(self, config):
+            self.done = False
+            super().__init__(config)
+
+        def run(self, parsed_args):
+            self.done = True
+
+    tc = TestClass("config")
+    tc.run([])
+    assert tc.done
+
+
+def test_basecommand_run_mandatory():
+    """BaseCommand subclasses must override run."""
+
+    class TestClass(BaseCommand):
+        help_msg = "help message"
+        name = "test"
+        overview = "fake overview"
+
+    tc = TestClass("config")
+    with pytest.raises(NotImplementedError):
+        tc.run([])
+
+
+def test_basecommand_mandatory_attribute_name():
+    """BaseCommand subclasses must override the name attribute."""
+
+    class TestClass(BaseCommand):
+        help_msg = "help message"
+        overview = "fake overview"
+
+    with pytest.raises(TypeError):
+        TestClass("config")
```

Comment:
Here and below are the new tests to check that attributes are mandatory in subclasses.